h2olog/misc/gen-quic-bpf.py

```diff
@@ -257,10 +270,75 @@ def prepare_context(h2o_dir):
 
   return context
 
+def build_bpf_header_generator():
+  generator = r"""
+#define GEN_FIELD_INFO(type, field, name) gen_field_info(#type, #field, &((type *)NULL)->field, name)
 
-def generate_cplusplus(context, output_file):
-  h2o_dir = context["h2o_dir"]
+#define DEFINE_RESOLVE_FUNC(field_type) \
+std::string gen_field_info(const char *struct_type, const char *field_name, const field_type *field_ptr, const char *name) \
+{ \
+    return do_resolve(struct_type, field_name, #field_type, field_ptr, name); \
+}
+
+template <typename FieldType>
+static std::string do_resolve(const char *struct_type, const char *field_name, const char *field_type, const FieldType *field_ptr, const char *name) {
+    char buff[1024] = {};
+    char *b = buff;
+    size_t s = sizeof(buff);
+
+    b += snprintf(b, s, "/* %s (%s#%s) */\n", name, struct_type, field_name);
+    s -= b - buff;
+    b += snprintf(b, s, "#define offsetof_%s %zd\n", name, (const char *)field_ptr - (const char *)NULL);
+    s -= b - buff;
+    b += snprintf(b, s, "#define typeof_%s %s\n", name, field_type);
+    s -= b - buff;
+    b += snprintf(b, s, "#define get_%s(st) *((const %s *) ((const char*)st + offsetof_%s))\n", name, field_type, name);
+    s -= b - buff;
+    b += snprintf(b, s, "\n");
+
+    return std::string(buff, b - buff);
+}
+
+DEFINE_RESOLVE_FUNC(int32_t);
+DEFINE_RESOLVE_FUNC(uint32_t);
+DEFINE_RESOLVE_FUNC(int64_t);
+DEFINE_RESOLVE_FUNC(uint64_t);
 
+static std::string gen_quic_bpf_header() {
+  std::string bpf;
+"""
+
+  for st_name, st_fields in struct_map.items():
+    # It limits a size of structs to 128
+    # This is because the BPF stack has a limit to 512 bytes.
+    generator += r"""
+  bpf += "#define sizeof_%s " + std::to_string(std::min<size_t>(sizeof(struct %s), 128)) + "\n";
+""" % (st_name, st_name)
+
+    for st_field_access, st_field_name_alias in st_fields:
+      name = "%s__%s" % (st_name, st_field_name_alias or st_field_access)
+      generator += """  bpf += GEN_FIELD_INFO(struct %s, %s, "%s");\n""" % (st_name, st_field_access, name)
+
+  generator += r"""
+  return bpf;
+}
+"""
+  return generator
+
+def build_typedef_for_cplusplus():
+  typedef = st_quicly_conn_t_def
+
+  # CAUTION: it depends on a GCC/Clang compiler extension: typeof()
+  i = 0
+  for st_name, st_fields in struct_map.items():
+    for st_field_access, st_field_name_alias in st_fields:
+      i += 1
+      typedef += """static struct %s s%d; // placeholder for typeof()\n""" % (st_name, i)
```

Comment:
I think we can get rid of this line. The type can be obtained as `typeof(((type *)NULL)->field)`.

Reply:
Actually! The same technique is used to calculate offsets.

Fixed 9a36c0c398c0e4459e7f710479ebac2f9e21fd5b